WIP: Add async SSR data fetching

LawJolla opened the pull request:
This is my hack, for lack of a better term, to allow data fetching in gatsby-ssr.  

**Code**: I used the async function from api-browser-ssr and then wrapped static-entry in a `then` after the promise-fied `replaceRenderer`.  
 
**Purpose**: This allows my application to walk my client's tree (`bodyComponent`), look for queries, and run those at compile time.  Here's my gatsby-ssr for reference.

https://gist.github.com/LawJolla/5fa01d54c64f1b17eea498b68068acab

You can read more about my use case here:  https://medium.com/@dwalsh.sdlr/gatsby-apollo-graphcool-netlify-the-webs-promised-land-6dd510efbd72

My application is an ecommerce site.  gatsby-node runs queries to determine the pages and gatsby-ssr runs the client queries to determine the data for those pages.  Then when Apollo Client loads, its store hydrates and all statically generated data is now controlled by the Apollo store, including new queries and web socket subscriptions.

Huge win!

KyleAMathews commented on packages/gatsby/cache-dir/static-entry.js:
```diff
@@ -99,117 +99,120 @@ module.exports = (locals, callback) => {
-      key="webpack-manifest"
-      dangerouslySetInnerHTML={{
-        __html: `
+      .then(() => {
```
Could you use async/await here?

LawJolla commented:
I like it!  Thanks for the comments and ideas, makes sense. Let me see if I can do this without blowing up the entire repo.  🥇  Hopefully I'll have up later today, this weekend at the latest.

KyleAMathews commented:
Deploy preview failed.

Built with commit 3779a1965f048dca5e6989710ff28e2f49e0ca08

https://app.netlify.com/sites/using-glamor/deploys/59e10a8b0b79b7400d06284e

KyleAMathews commented:
Deploy preview failed.

Built with commit 2a2c5055f6d72d8c7bbe9eeff040336f0e123e6a

https://app.netlify.com/sites/using-glamor/deploys/59e10b06a6188f2f0db30317

LawJolla commented:
I'm a solo developer, so I'm terrible with git.  I had a merge conflict and I think I fixed it..?  I tried to rebase after I sent the updated PR, but I just blew up my local repo more.  

It looks like the code is right.  If not, please let me know...  To save my sanity, it would be easier for me to close this PR, I'll fork it again and reapply my changes and submit a new PR.

Onto the substance... I got static-entry to work for both asyncSSR = true and asyncSSR = false pages!  I also updated the snapshot tests to add asyncSSR.

On the api changes, I renamed api-runner-ssr functions to `apiRunnerSync` and `apiRunnerAsync`.  I thought it made the code clearer.  Whatever you think.

The other change is a little controversial.  For `replaceRenderer`, I added `page: Object.assign({}, page)`.  Having the page object was a huge help to me to troubleshoot gatsby-ssr.  When gatsby-ssr threw errors, I could figure out my asyncSSR status and path.  

Please let me know what else I can do.  I'm happy that I was able to implement the feature!

LawJolla commented:
I'm not sure if this PR is failing the deploy build.  Above it says it is, but below it says that its passed.  

It also says it failed the windows build 
`- yarn (exited 1603) - Error while running 'C:\ProgramData\chocolatey\lib\yarn\tools\chocolateyinstall.ps1'`

If anyone can help me understand where to start and the changes needed, I'm happy to make them!

KyleAMathews commented:
@LawJolla that was fixed fairly recently so once you've resolved the yarn.lock error you'll be fine.

BTW, sorry for the slow merging of this — it's a big change and I haven't had time to do a proper think of its implications :-(

LawJolla commented:
I guess I should have tried to merge instead of rebase. 😪  What a mess.  I'm so bad at git.

If this is ok, great.  If not, please feel free to close it and I'll PR a new fork.

It looks like the builds/tests are failing for reasons unrelated to my commits.

KyleAMathews commented:
Yeah, easiest thing to do is click the "Update branch" button at the bottom of PRs. That being said, what you did is fine too and doesn't hurt anything :-)

![screen shot 2017-11-02 at 9 05 04 am](https://user-images.githubusercontent.com/71047/32336612-fd58e62c-bfac-11e7-8331-a02616e3505f.png)

sedubois commented:
Why is the deploy preview called “using drupal”?

KyleAMathews commented:
@sedubois we build all the example sites on every PR including using-drupal. As integration tests basically.

imjared commented:
@LawJolla this looks like something I could really make use of. Is there anything I could help with to push this along and get it merged?

LawJolla commented:
@imjared Huge thanks for the offer and interest!

Before I destroyed this PR with my rebase mess 😄 , @KyleAMathews mentioned that the PR is a big change and he needed more time to consider its consequences.  But more interest in this PR helps!

samwiseduzer commented:
I'm VERY interested in this and would be interested in contributing - I'll be anxiously returning to see the current status.

LawJolla commented:
Also, if anyone has feedback, that would be great.  You can clone my repo at:  `git clone git@github.com:LawJolla/gatsby.git`

Branch is topics/ssr-async

Then use the excellent Gatsby Dev CLI.  Point it at the cloned repo and then run it in your project.  https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-dev-cli

The only API change is the page object how has an `asyncSSR` property.  Just set it to true, `page.asyncSSR = true`, and `createPage`!

KyleAMathews commented:
@LawJolla if we added support for "schema stitching" would that solve your problem? I.e. instead of directly querying remote GraphQL APIs, we would simply stitch the remote schema into the Gatsby generated one.

LawJolla commented:
Hi @KyleAMathews !  

I haven't worked with schema stitching beyond reading a few articles, but I don't think it'll solve the problem.  My problem is query duplication. 

Allow me to carefully lay out what I'm doing and why I'm doing.  I'm surprised that my use case hasn't been a bigger pain point to others.  That makes me think either I've gone down the wrong path or "live data" sites simply aren't on the static site train (yet).  

I own a car dealership and developed my own software.  Version 2 is a complete rewrite and will be multi tenant / SaaS app for other dealerships.  I have a lot of changing data and quite dynamic product pages.  However, most of that data is "below the fold."  So while I want the data to be fresh, I also don't need to trigger rebuilds for every change.

For the vehicle page(s), I'll have the following client query...
```
query {
      vehicles {
          id
          year
          make
          offers
          ...
      }
  }
```

To get there, my sourceNodes query is:

```
{
  allDealerships(filter:{ name:"Wheel Kinetics"}) {
    vehicles(filter: {isSold: false}) {
       id
    }
  } 
}
```

And createPages ...

```
  {
          allVehiclePage(limit: 100) {
            edges {
              node {
                field
              }
            }
          }
        }
```

I've delegated the routing to Gatsby and data to the client / Apollo.  

At runtime,  the compile time Apollo query statically renders.  When the JS bundle loads, Apollo's 'cache-and-network` policy uses the cache (hydrated from the static build), but then reruns queries at runtime and updates the cache with new data.  (With Apollo, it has the additional benefit of instantly showing the new idea if it encounters the same field again).

You can see an example here with the truck's mileage.
https://frame-maker-catcher-57758.netlify.com

I'm not sure how to replicate that in Gatsby without great difficulty.  Apollo (and even more so Relay) encourage fragments of queries.  To implement this without the PR, I think I would need to walk the component tree in `gatsby-node` (but I don't think the tree is complete until bodyComponent in SSR)  and then somehow initialize Apollo on top of the rendered data at runtime. 

I want to delegate my client data needs to Apollo.  With this PR, I get the same experience as building a client side app by allowing Apollo to walk bodyComponent, parse its queries, and run them via an async bodyComponent function.

LawJolla commented:
@sedubois "sitching data".. I like that!  That's exactly what it is.

sedubois commented:
GraphQL data stitching has definitely become a buzzword recently. Any chances that Gatsby could migrate to the Apollo 2 client which allows to do that? @kyleamathews

KyleAMathews commented:
No definitely have heard from a number of people that they want the ability to query "live" data once a user is browsing in the client.

I'd really like to do it in a pretty seamless fashion though if possible — that Gatsby could automatically know that certain fields are from an external GraphQL server and automatically start querying those live once in the browser through Apollo. Then we could automatically take advantage of Gatsby's smart prefetching to speed up browsing in the site. I noticed for example that your example site was a bit laggy clicking on the links because we had to wait for data to return. With Gatsby's automatic prefetching that'd be eliminated.

I also want to retain the ability in the future to do incremental builds — Gatsby, with the current setup, can track exactly what data is used on every page — this is what will enable incremental builds as we'll be able to say, "oh, node 714 changed. That node is used on these 4 pages so we'll rebuild those".

That's why I think stitching is the best approach here as we'll retain the ability to track dependencies between data and pages & it'll be easy for Gatsby to know also what data is being fetched from an external API and automatically start fetching that data live in the client.

We'll create a plugin e.g. gatsby-plugin-apollo which will make this happen again, automatically for people.

With the world move strongly towards GraphQL, this sets us up really nicely. You can wrap old skool REST APIs with Gatsby source plugins and then "stitch" in GraphQL APIs just by adding them to your config.

KyleAMathews commented:
I'm very confident we can do this. My main concern though is I don't want to block anyone from using Gatsby so wondering if we should still add this API or something similar as a general purpose escape hatch.

I might have a client soon that needs this so hopefully can start working on this in the new year.

atrauzzi commented:
Any possible connection here to my old ticket about any convention for configuring the built site at run time?

cc. https://github.com/gatsbyjs/gatsby/issues/1527

LawJolla commented:
Thanks for laying out the concerns!  

Maybe you can give me some guidance for other areas of investigation.  I want Gatsby to fetch Apollo client queries.  The only way I know to get React components for a given page is from `bodyComponent`.  Then let Apollo's `getDataFromTree` function walk it and fetch the queries.

`getDataFromTree` mutates the previously created `client` store.  The data from that store can be extracted with `getInitialState` and merged with Gatsby's data for incremental builds.

It seems like a hard problem because of the life cycle.  Gatsby wants to keep incremental build data to gatsby-node, but bodyComponent is called after in ssr.  Could it make sense to compose the React tree in gatsby-node?  Perhaps return the tree from `createPage` if possible, e.g.  (I didn't review the docs before this, so this is just pseudo code)

```
// gatsby-node createPages()

const apolloState = new ApolloClient({ ... })

const { reactComponents, pageId } = createPage({
              ...
})

const  _ = await getDataFromTree(reactComponents)

modifyPageContext(pageId, apolloState, `apolloState`)
```

```
// gatsby-ssr  replaceRenderer()
// make `page` available to replaceRenderer

 const app = (
<ApolloProvider client={page.context.apolloState}>
      {bodyComponent}
</ApolloProvider>
)

const body = replaceBodyHTMLString(app)
setHeadComponents([page.context.apolloState.getInitialState()])

return
```

KyleAMathews commented:
@LawJolla we wouldn't use apollo on the backend — nothing would change there. We'd just wire up things so the Apollo client gets the data on the frontend. We'd continue to control prefetching data just now it'd be through Apollo and hitting the live server.

LawJolla commented:
I'm not suggesting to use it on the backend.  It should obviously be agnostic about GraphQL client choice (or none at all).  I was just trying to show a potential means of integrating it / how to make it happen.  

When you get some time in the New Year and have an implementation idea, I'm happy to help!

ViliamKopecky commented:
@KyleAMathews asked me to share my use case here as I was asking for option to use async `getInitialProps` next to exported graphql `pageQuery`.

The original mission is to enable fast previews, when editing WordPress posts. And by thinking about solutions I had two ideas. But these ideas dont actually solve the previews issue 😁

### afterPageQuery

Possibility to define some callback to modify graphql result after `pageQuery`. The final result would be the static payload for pages.

```jsx
// src/templates/page.js

export const pageQuery = graphql`...`;
export const afterPageQuery = async (graphqlResult, pageContext) => {
  // do some magic, even async
  graphqlResult.customized = true;
  return graphqlResult;
};
```

*For example I would like to transform the resulting structures of `edges[n].node` to regular arrays. That way the resulting page component wouldnt need to do it every time.*

### getInitialProps

Then I was thinking - why not just give the developer the option to have custom provider for the static data. If they know what they are doing. When you are ok with not using graphql. (naming taken from next.js).

```jsx
// src/templates/other-page.js

export const getInitialProps = async (defaultPropsFromGatsby, pageContext) => {
  // do some magic, even async
  defaultPropsFromGatsby.customized = await someAsyncDataLoader(pageContext.id);
  return defaultPropsFromGatsby;
};
```

### solving WordPress previews

Things above wouldnt help in solving the previews issue. But my thinking was, that we could prepare the static data for pages on the WordPress server. Then `getInitialProps` would get it from WP. And we would create some component decorator to check if page is in preview mode and if so it would just swap original props with static data with new ones, that would be generated by wordpress with values from preview.

sedubois commented:
My first thought was that it would be great not to increase the API surface, and why bypass the GrahQL layer? It sounds like the strong point of Gatsby and bypassing it looks like a workaround rather than a fix?

GerritVK commented:
Anything new on this discussion, I am very interested in this topic for a handful of projects?

LawJolla commented:
It looks like Kyle and Apollo are working together for a more comprehensive solution.  Kyle has the laudable goal that Gatsby tracks which pages need rebuilding, rather than an entire site rebuild.  With my PR, Gatsby can't know about the Apollo data.  (And I don't know how to implement it without more Gatsby API changes, like building the `bodyComponent` in `createPage`)

LawJolla commented:
Hi @KyleAMathews! 

When you get a moment, I'm hoping to get your thoughts on the following.  I've gotten a few comment on my approach that I'm "misusing" Gatsby when I really should be doing SSR.  I have my own feelings, but what do you think -- should this style of "app" be relegated to SSR?

lionkeng commented:
@KyleAMathews is there ongoing work on gatsby-plugin-apollo as mentioned by you in one of your previous comments? Thx.

shwanton commented:
I have a e-commerce site that is using Gatsby to render all the product pages. I'm using Apollo on the client and during runtime I fetch dynamic data like current inventory. I also need client-side search using `apollo-link-state` (similar to redux).

Ideally, I'd like to pre-poulate my Apollo store with all my products so I can have client side search without making an additional query for all items. I'm already querying all the data to build the nodes and pages in Gatsby so it would be great to use those queries to hydrate local state.

A temporary solution I've come up with is having a pre-build step that will create a new `ApolloClient` with my `FetchAll` query in a dummy Component, run Apollo's `getDataFromTree` and copy the stringified state data to a text file. When I build the Gatsby site, I just read from that text file and hydrate local state in `gatsby-ssr`

```
const client = new ApolloClient({
  cache,
  ssrMode: true,
  link,
})

const App = (
  <ApolloProvider client={client}>
    <AppDataLoader />
  </ApolloProvider>
)

getDataFromTree(App)
  .then(() => {
    const state = client.cache.extract()
    fs.writeFile('./public/client.json', JSON.stringify(state), () => {})
  })

import state from './tmp/state.txt'
__html: `window.__APOLLO_STATE__=${state};`,
```

It would be great to have an out of the box solution for this since the data is all there during the build process.

LawJolla commented:
@shwanton it sounds like we're doing the same thing... I like your approach because it end runs Gatsby entirely where mine requires a PR/fork.

What are you doing for your Gatsby queries?  My Gatsby queries are just enough to generate routes, e.g. find the products, create the slugs for the products, and then make the nodes.  The data for those product pages comes from the Apollo tree walk.

KyleAMathews commented:
>is there ongoing work on gatsby-plugin-apollo

Not ongoing but it's planned either for the v2 release or during the v2 release cycle.

shwanton commented:
@LawJolla I'm using data from a small graphql server I build over the square rest api.

I use a `graphql` query for creating my Gatsby nodes and then use those nodes for creating my product pages. I have some client only queries (inventory, checkout) that I don't want to be part of the SSR render step since the data is very dynamic. I don't necessarily want those queries as part of my SSR Apollo crawl.

In my `<AppDataLoader />` component, I just use the same `graphql` query that I ran to create my Gatsby nodes. 
```
const AppDataLoader = () => {
  return <div>DataLoader!</div>
}

export default graphql(gql`
  query FetchAll {
    brands {
      id
      ...brandDetails
      items {
        id
        itemDetails
      }
    }
  }
`)(AppDataLoader)
```

My pre-build step will use this as the only child component with the same Apollo client as my app.
```
const App = (
  <ApolloProvider client={client}>
    <AppDataLoader />
  </ApolloProvider>
)
```

The weird part is my page templates are querying the Gatsby data by `slug` (to create pages) instead of by `id` which the api is using. This would make it tricky to just share these queries. However all the data has already been queried during the node creation, so I'm sure it's just a matter of figuring out how to write the correct query for it.

A thing it took me a bit to figure out was `apollo-link-state` for combining the data from graqhql with my local state data (Cart, UI State etc...). As long as you setup the `apollo-client` correctly, you should be able to hydrate local `apollo` state with the correct defaults.

I was imagining this could all run as part of a Docker image that could run the pre-build step so it's available to `gatsby build`. 

I'll try to extract this all into a working demo.

MathieuDelafosse commented:
@LawJolla Thank you for this awesome work.
@shwanton Too ! Did you find the time to create a working demo of your code ? Sounds promising since I'm in the same use case where I need to create products pages where some data are highly dynamic (as the price) and need to fetch them "live" client side.
Our implementation is quite recent and we are still struggling on details but the read of this PR have been highly instructive.
Finally, since both of you have been in touch with @KyleAMathews do we have additional informations about `gatsby-apollo-plugin` ?

LawJolla commented:
Hi @MathieuDelafosse ,

Thanks for the kind words and interest.  I haven't heard anything on this PR in some time and I think it's still here just for a discussion thread.  

I'm getting the sense that our dynamic page use case is not where Gatsby wants to go and we'd be better off with an SSR package like NextJS.

But there's still a place for moderately dynamic Apollo sites using Gatsby that would benefit from a Gatsby-SSR query/compile step.

pieh commented:
> I haven't heard anything on this PR in some time and I think it's still here just for a discussion thread.

@LawJolla In order to clean up the PR list I'll close hanging PRs like this over the next few days. Can you capture problem this PR is/was trying to solve as Issue and add link to discussion in this PR there so it doesn't get lost?

fresh5447 commented:
Is this PR still active? I'd love to see it get merged.

m-allanson commented:
I'm going to close this as it has a lot of conflicts that will need tidying up. Would be happy to see discussion about this carried on in a new PR or issue 👍

LawJolla commented:
@m-allanson this PR was mergable for months but ignored by Kyle. Ruined my opinion of the Gatsby project, so I switched my company to NextJS and have no further interest in this PR or any contribution. 

Good luck getting your free socks everyone!

KyleAMathews commented:
Hey @LawJolla. Sorry about this. I should have closed this a long time ago — I just dithered because in many ways your PR _is_ a really good idea and you put a ton of time into it and really pleased about it with your blog post, etc.

Also what you tackled _is_ a real problem with Gatsby that we still don’t have a great solution for so I kept thinking that perhaps I’d warm to your work.

But that was unkind and I should have been decisive.

I'm really sorry about that as I'm sure it hurt personally and caused troubles for you deciding what technical direction to take. We'll do better in the future.

sedubois commented:
I’m also looking to any followup to this PR with great interest. I also think @lawjolla did an amazing job and I hope the ideas presented here will soon bear fruit in another form. Meanwhile I wish all the best to the Gastby team for the v2 release!

LawJolla commented:
Thanks @KyleAMathews I really appreciate that.

i8ramin commented:
Hi, is there any more progress on this topic? We also have some requirements around fetching data during the SSR process and being able to use async/await with the `replaceRenderer` function would be really handy.

leonardodino commented:
I'm sorry for bumping this up again, but is something along this lines on the roadmap?
Would another PR along this lines be considered @KyleAMathews?

---

This would seriously improve some use-cases.

I could make a gatsby plugin for ssr with [`graphql-hooks`](https://github.com/nearform/graphql-hooks/) with only a couple of lines of code.

---

I'm using an external `graphql` api for comments in a blog.
I want them to be the static html render (for SEO) reasons and also to be live when someone posts.
Right now the logic would have to be duplicated in a couple of places. 😕

BrunnerLivio commented:
The StencilJS community is also really interested in the PR! Stencil offers a [`hydrate.renderToString()`](https://stenciljs.com/docs/hydrate-app) function, which is async since the components Lifecycle Events could be async. I would love to see an alternative PR!

Jefftopia commented:
My team is also adopting Stencil based WebComponents. The extent to which we use Gatsby going forward for static sites heavily depends on support for async SSR. My team is pretty big, with a lot of Gatsby fans. It would be a shame to go elsewhere due to this.

jonearley commented:
@BrunnerLivio, @Jefftopia -- Here's a way to SSR Stencil components without needing to use Gatsby's `replaceRenderer`, but by processing all the built html files.

https://www.npmjs.com/package/gatsby-plugin-stencil

Jefftopia commented:
@jonearley Brilliant, thanks so much!
